Lab2 bubble sort

sahidakhtar1 opened the pull request:
Fixes #43 

#### Changes proposed in this PR:

- `bubbleSort()` to sort a list of `int`
- UnitTest - Use of table driven testing.

Commit:
Add method to perform bubble sort on a list of `int`.

Commit:
Add Unit test cases to test bubble sort using table driven testing.

Commit:
Create folder structure 02_bubble/USERNAME

Commit:
Use string literals directly instead of strconv.Quote

Commit:
Use direct string literals instead of `strconv.Quote`

Commit:
Merge branch '02_bubble' of https://github.com/sahidakhtar1/go-training into 02_bubble

anzdaddy commented on 02_bubble/sahidakhtar1/main.go:
```diff
@@ -0,0 +1,29 @@
+		}
+		count = swappedIndex
+	}
+}
```
Can't mutate the original. Make a copy and return it. Also write a test to check that the original isn't mutated.

juliaogris replied:
That means make a `copy()` of list and modify it.

sahidakhtar1 replied:
`bubbleSort ()` will return new list without modifying the input. Also added Unit test case to check input is not modified `TestBubbleSortInput`.

juliaogris commented on 02_bubble/sahidakhtar1/main.go:
```diff
@@ -0,0 +1,29 @@
+func main() {
+	list := []int{3, 2, 1, 5}
+	bubbleSort(list)
+	fmt.Fprintln(out, "After sorting = ", list)
```
That's not the requested output.

sahidakhtar1 replied:
Changes done.

juliaogris commented on 02_bubble/sahidakhtar1/main_test.go:
```diff
@@ -0,0 +1,41 @@
+	{[]int{3, 2, 1, 5}, []int{1, 2, 3, 5}},
+	{[]int{3, 2, 1, 5, -1}, []int{-1, 1, 2, 3, 5}},
+	{[]int{3, 2, 1, 5, -1, 3}, []int{-1, 1, 2, 3, 3, 5}},
+	{[]int{3, 2, 1, 5, -1, 3, 0}, []int{-1, 0, 1, 2, 3, 3, 5}},
```
Nice!
Add corner cases: empty slice, one value slice, slice with duplicated elements, sorted slice, reverse sorted slice.

sahidakhtar1 replied:
Added the corner cases.

camscale replied:
Just missing the duplicated elements test case. With that done, it should all be good.

sahidakhtar1 replied:
@camh-anz,its there {[]int{3, 2, 1, 5, -1, 3}, []int{-1, 1, 2, 3, 3, 5}} 3 is duplicate here.

camscale replied:
sorry, i missed that. all good.

juliaogris commented on 02_bubble/sahidakhtar1/main_test.go:
```diff
@@ -0,0 +1,41 @@
+	r.Equalf(expected, actual, "Unexpected output in main()")
+}
+
+var testData = []struct {
```
You should probably define this data as local variable to `TestBubbleSort`

sahidakhtar1 replied:
Added `TestBubbleSortInput` which uses the same inout so not making any changes to it.

juliaogris commented on 02_bubble/sahidakhtar1/main_test.go:
```diff
@@ -0,0 +1,41 @@
+	{[]int{3, 2, 1, 5, -1, 3, 0}, []int{-1, 0, 1, 2, 3, 3, 5}},
+}
+
+func TestBubbleSourt(t *testing.T) {
```
`s/TestBubbleSourt/TestBubbleSort `

sahidakhtar1 replied:
Corrected the typo.

juliaogris reviewed (commented):
Click `Update branch` and locally `git pull origin lab2` before adding new commits.

Commit:
Change to address review comments.
	- bubbleSort() will return new sorted list without modifying the original list.
	- Test case to test the input value is not modified

Commit:
Fix GoLangCI issue.

Commit:
Change to update test case to check input values to bubbleSort()

sahidakhtar1 committed:
Merge branch 'master' into 02_bubble

camscale commented on 02_bubble/sahidakhtar1/main_test.go:
```diff
@@ -0,0 +1,55 @@
+		input := make([]int, len(tt.in))
+		copy(input, tt.in)
+		bubbleSort(tt.in)
+		r.Equalf(input, tt.in, "Test case fails")
```
The string would be better saying something "sort input should not be mutated" or something that describes the test failure more specifically. The test harness will already tell you that the test case failed."

camscale commented on 02_bubble/sahidakhtar1/main_test.go:
```diff
@@ -0,0 +1,55 @@
+	r := require.New(t)
+	for _, tt := range testData {
+		result := bubbleSort(tt.in)
+		r.Equalf(tt.out, result, "Test case fails")
```
The string would be better saying something like "sorted output not correct" or something like that, as it describes the failure that happened. The test harness will already tell you that the test case failed.

camscale reviewed (approved):
LGTM. Ready to merge. I've added a couple of comments, but that is for you to keep in mind as you move through the next labs. I wouldn't worry about updating this PR to fix up these small things - you'll have plenty of opportunity as you go through the labs to get to these sort of details.

Thanks.

camscale commented:
I forgot to say, excellent commit titles. Keep doing that!
